### PR TITLE
Implement IOWebSocketChannel as a WebSocketAdapterWebSocket subclass

### PR DESCRIPTION
**Pull request by brianquinlan**
See [Simplify the web_socket_channel implementation](https://github.com/dart-lang/web_socket_channel/issues/340)

---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: lib/web_socket_adapter_web_socket_channel.dart [hunk 1]
```diff
@@ -66,14 +66,19 @@ class WebSocketAdapterWebSocketChannel extends StreamChannelMixin
...
+      WebSocketAdapterWebSocketChannel(
           WebSocket.connect(url, protocols: protocols));
 
-  // Create a [WebSocketWebSocketChannelAdapter] from an existing [WebSocket].
-  factory WebSocketAdapterWebSocketChannel.fromWebSocket(WebSocket webSocket) =>
-      WebSocketAdapterWebSocketChannel._(Future.value(webSocket));
+  // Construct a [WebSocketWebSocketChannelAdapter] from an existing
+  // [WebSocket].
+  WebSocketAdapterWebSocketChannel(FutureOr<WebSocket> webSocket) {
+    late Future<WebSocket> webSocketFuture;
```
Comment (natebosch) [inline lib/web_socket_adapter_web_socket_channel.dart:75]: Is this `late` necessary? I'd expect flow analysis to see this is definitely assigned.

File: lib/web_socket_adapter_web_socket_channel.dart [hunk 2]
```diff
@@ -113,7 +118,16 @@ class WebSocketAdapterWebSocketChannel extends StreamChannelMixin
       _protocol = webSocket.protocol;
       _readyCompleter.complete();
     }, onError: (Object e) {
-      _readyCompleter.completeError(WebSocketChannelException.from(e));
+      late Exception error;
```
Comment (natebosch) [inline lib/web_socket_adapter_web_socket_channel.dart:121]: Is this `late` necessary?

**Review by natebosch — APPROVED** [review]

File: lib/web_socket_adapter_web_socket_channel.dart (hunk 1 above)
Comment (brianquinlan) [reply lib/web_socket_adapter_web_socket_channel.dart]: Nope, thanks!

File: lib/web_socket_adapter_web_socket_channel.dart (hunk 2 above)
Comment (brianquinlan) [reply lib/web_socket_adapter_web_socket_channel.dart]: Nope, thanks!

**Pull request merged**